ath79: enable LNA (Low Noise Amplifier) for TPLink CPE210/220/510/520

thuehn opened the pull request:
The hardware has a build-in LNA. This patch activates those in the receiving
chain and improves the rx sensitivity by about 20dB.

Tested on CPE210v1, CPE510 v1, v2 & v3.

Signed-off-by: Thomas Huehn <thomas.huehn@hs-nordhausen.de>

adschm commented:
For the two-ports this is already enabled in ar9344_tplink_cpe_2port.dtsi
Are you sure it's also required for the one-ports?

@apcameron Can you comment?

thuehn commented:
Hi,

I'am quite sure ... as I'am in the process of merging our mesh routers from ar71xx to ath79 and the ath79 for CPE510v1,v2,v3 and CPE201v1 had shown a poor receiving performance with actual trunk... after enabling LNA it looks much better (rx sensivity is significantly increased)

Greetings Thomas

adschm commented:
Well, for CPE210/510 v1 the nodes you introduce are redundant. So, if you observe a performance change there, your test setup is broken/wrong.

thuehn commented:
ah that was not obvious for as they have their individual dts file and when I choose the cpe510v1 traget I do not get a CPE210 image file. I tested latest trunk from download folder against my own build with the patch ... so I think my test-setup is reasonable nor ?

apcameron commented:
I suggest asking @robimarko for input on this.

adschm commented:
> ah that was not obvious for as they have their individual dts file and when I choose the cpe510v1 traget I do not get a CPE210 image file. I tested latest trunk from download folder against my own build with the patch ... so I think my test-setup is reasonable nor ?

Again: Your patch does not change anything for CPE210 v1 and CPE510v1 (2-port devices). The gpio-hogs are already present for these.

The only difference is for CPE510 v2/v3 (1-port devices), where the gpio-hogs have not been present and your patch adds them.

So, if you measure a difference between master and your build for CPE210v1/CPE510v1, it's definitely not due to the patch proposed here. There might be differences between ar71xx and ath79 in general, though.

thuehn commented:
Hi Adrian,

I miss understood your comment, so you are refering to the dts file ar9344_tplink_cpe_2port.dtsi, where the LNAs already got enabled... I obviously missed that. So would you suggest to add the LNA enable part for the 1-port devices in ar9344_tplink_cpe.dtsi ?
Greetings Thomas

musashino205 commented:
https://github.com/openwrt/openwrt/blob/dfa357a3def512c13f22371d24138b6e8093be18/target/linux/ath79/dts/ar9344_tplink_cpe_2port.dtsi#L47-L61

adschm commented:
> I miss understood your comment, so you are refering to the dts file ar9344_tplink_cpe_2port.dtsi, where the LNAs already got enabled... I obviously missed that. So would you suggest to add the LNA enable part for the 1-port devices in ar9344_tplink_cpe.dtsi ?

That obviously depends on whether they need them ...

_If_ they need these definitions as well, then one should move the gpio-hogs block from ar9344_tplink_cpe_2port.dtsi to ar9344_tplink_cpe.dtsi.

thuehn commented:
Ah ... sorry for me not having a good knowledge yet about that fact that dts files structures can be expressed in such a hierarchical manner .. cool... so I can update the patch and I have tested the LNA feature on one-port devices CPE510v2 & v3 with good results in terms of increases rx-sensivity. Just check my stock of devices and I have a CPE210v3 where I can test it as well and report back. I do not have a CPE210v2 around. What do you think is sufficient to aplly LNS for all 1-port devices ?

Greetings Thomas

adschm commented:
>  I have a CPE210v3 where I can test it as well and report back. I do not have a CPE210v2 around. What do you think is sufficient to aplly LNA for all 1-port devices ?

CPE210(!) v2/v3 are qca9533 SoC, while CPE510 v2/v3 are ar9344 SoC like the CPE210 v1 and CPE510 v1.

So, these will certainly be quite different and be handled as such.

I do have a CPE210v2 myself BTW.

> Ah ... sorry for me not having a good knowledge yet about that fact that dts files structures can be expressed in such a hierarchical manner .. cool... 

Just look at the DTS files. There are includes at the top. Essentially, the parser just inserts these includes into the DTS file, creating a big final file with all the definitions.

robimarko commented:
As far as I remember only AR934x based ones have LNA-s that need to be enabled while QCA953x ones dont have those.

adschm commented:
@robimarko So, we should enable for _all_ ar934x TP-Link CPE? And the fact that it's currently only for the two-ports is probably a mistake/has been overlooked?

apcameron commented:
@robimarko Do the One Port devices have one LNA or two? If the One Port device only has one that needs enabling we would need to take that into account.

robimarko commented:
LNA-s are for RF, they have no relation to LAN ports.
As far as I know, there should be two on all boards that have them.

thuehn commented:
I updated my PR to have the LNAs enabled on all 1-port CPE devices.

adschm commented:
Your commit has formal issues, i.e. the title is too long.

Apart from that, please don't create an additional setting for 1-port devices, but move the block from two-port devices to the shared ar9344_tplink_cpe.dtsi as requested above.

thuehn commented:
Hi Adrian,

Thx for your review, I have fixed the commit message and changed the patch to have LNA enabled on all CPEs.

adschm commented:
> ath79: enable LNA for all TPLink CPE devices

Actually it's only all ar9344 TP-Link CPE devices ...

thuehn commented:
Hi Adrian,
Thx for your comment, I fixed the commit message to make clear the path is for ar9344.

Greetins Thomas

adschm commented:
@robimarko Can I use your Acked-by on this one?

thuehn commented:
sure

robimarko commented on target/linux/ath79/dts/ar9344_tplink_cpe.dtsi:
```diff
@@ -114,3 +114,19 @@
+&gpio {
+       gpio_ext_lna0 {
+               gpio-hog;
+               gpios = <18 0>;
```
Use macros for GPIO logic level.
GPIO_ACTIVE_HIGH

robimarko commented on target/linux/ath79/dts/ar9344_tplink_cpe.dtsi:
```diff
@@ -114,3 +114,19 @@
+
+       gpio_ext_lna1 {
+               gpio-hog;
+               gpios = <19 0>;
```
Same as above

robimarko commented:
@adschm I just have remarks about not using macros for GPIO_ACTIVE_HIGH.
Other than that, feel free to add my ACK.

adschm commented:
@thuehn Please apply the changes suggested by robimarko and use tab indent (found by checkpatch.pl).

thuehn commented:
Hi .. just updated the patch with the comments from robimarko.
Greetings Thomas

adschm commented:
Tabs!

thuehn commented:
Hi Adrian,
thx for been so patient .. I've missed this one .. fixed my vim conf ... and the patch should be proper tab intended by now ;)
Greetings Thomas

blocktrron commented:
@thuehn 

> WARNING: From:/Signed-off-by: email address mismatch: 'From: Thomas Huehn <thomas@net.t-labs.tu-berlin.de>' != 'Signed-off-by: Thomas Huehn <thomas.huehn@hs-nordhausen.de>'

Please fix your Signed-off-by (or author E-Mail). Otherwise LGTM

thuehn commented:
Hiho Blocktrron... on the finishing line of my commit attempt... you have a valid point.
My email commit email setup does not match my current email address... I have changed my ~/.gitconfig .. but I have no glue how to validate my commit regarding this setting... I am just refresh my PR after updating my ~/.gitconfig.

dangowrt commented:
Moin @thuehn,
Good to see you here :)
I guess the only easy way to edit an existing git commit on that level (change things in the headers such as author, date, ...) is to export it using `git format-patch -1 ...`, then edit the resulting patch file, and then re-import it using `git am ...`.

blocktrron commented:
Do a `git commit --amend --reset-author --no-edit`

thuehn committed:
ath79: enable LNA for all ar9344 TPLink CPE devices

The TPLink CPE devices CPE210/CPE510 based on ar9344 have a build-in
Low Noise Amplifier on both of the 2x2 mimo rx chains.
This patch activates those two LNAs in the respective receiving chains
and hence improves the RX sensitivity by about 20dB.

Tested on CPE510 v2 & v3.

Signed-off-by: Thomas Huehn <thomas.huehn@hs-nordhausen.de>

thuehn commented:
... so ... I mananged to get my author and signed-off to be the same .. puhh

adschm commented:
Thanks, will be merged later today.
